x/ibc/06-solomachine/misbehaviour.go

```diff
@@ -0,0 +1,69 @@
+package solomachine
+
+import (
+	"bytes"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	clientexported "github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
+	clienttypes "github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+)
+
+// CheckMisbehaviourAndUpdateState determines whether or not the currently registered
+// public key signed over two different messages with the same sequence. If this is true
+// the client state is updated to a frozen status.
+func CheckMisbehaviourAndUpdateState(
+	clientState clientexported.ClientState,
+	_ clientexported.ConsensusState,
+	misbehaviour clientexported.Misbehaviour,
+) (clientexported.ClientState, error) {
+
+	// cast the interface to specific types before checking for misbehaviour
+	smClientState, ok := clientState.(ClientState)
+	if !ok {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "client state type %T is not solo machine", clientState)
+	}
+
+	if smClientState.IsFrozen() {
+		return nil, sdkerrors.Wrapf(clienttypes.ErrClientFrozen, "client is already frozen")
+	}
```

Comment:
this is not on the spec, I think the client should allow for a client to accept misbehavior if the submitted height is lower than the current frozen height. 

EDIT: ahh the client state doesn't have a frozen height, it's just a bool. Should we consider changing it @cwgoes ?

Reply:
Hmm - there isn't as much of a case for wanting to accept packets out of order, though I suppose it depends on the use-case. I'd like to solicit Agoric's opinion on this question - https://github.com/cosmos/ics/issues/427.

Reply:
see [`cosmos/ics#427(comment)`](https://github.com/cosmos/ics/issues/427#issuecomment-634101531) for conclusion - will use `bool` for `Frozen`